packages/react/src/jsx/ReactJSXElement.js

```diff
@@ -158,30 +158,81 @@ function defineRefPropWarningGetter(props, displayName) {
  * @internal
  */
 function ReactElement(type, key, ref, self, source, owner, props) {
-  const element = enableRefAsProp
-    ? {
-        // This tag allows us to uniquely identify this as a React Element
-        $$typeof: REACT_ELEMENT_TYPE,
-
-        // Built-in properties that belong on the element
-        type,
-        key,
-        props,
-
-        // Record the component responsible for creating this element.
-        _owner: owner,
-      }
-    : {
-        // Old behavior. When enableRefAsProp is off, `ref` is an extra field.
-        ref,
-
-        // Everything else is the same.
-        $$typeof: REACT_ELEMENT_TYPE,
-        type,
-        key,
-        props,
-        _owner: owner,
-      };
+  let element;
+  if (__DEV__ && enableRefAsProp) {
+    // In dev, make `ref` a non-enumerable property with a warning. It's non-
+    // enumerable so that test matchers and serializers don't access it and
+    // trigger the warning.
+    //
+    // `ref` will be removed from the element completely in a future release.
+    element = {
+      // This tag allows us to uniquely identify this as a React Element
+      $$typeof: REACT_ELEMENT_TYPE,
+
+      // Built-in properties that belong on the element
+      type,
+      key,
+
+      props,
+
+      // Record the component responsible for creating this element.
+      _owner: owner,
+    };
+    if (ref !== null) {
+      Object.defineProperty(element, 'ref', {
+        enumerable: false,
+        get() {
+          if (__DEV__) {
+            const componentName = getComponentNameFromType(type);
+            if (!didWarnAboutElementRef[componentName]) {
+              didWarnAboutElementRef[componentName] = true;
+              console.error(
+                'Accessing element.ref is no longer supported. ref is now a ' +
+                  'regular prop. It will be removed from the JSX Element ' +
+                  'type in a future release.',
+              );
+            }
+          }
+          return ref;
```

Comment:
Instead of closure per element it might be worth having a hoisted getter that returns `this.props.ref`.